libraries/libstratosphere/include/stratosphere/lm/detail/lm_detail_log_packet_transmitter.hpp

```diff
@@ -0,0 +1,122 @@
+/*
+ * Copyright (c) 2018-2020 Atmosphère-NX
+ *
+ * This program is free software; you can redistribute it and/or modify it
+ * under the terms and conditions of the GNU General Public License,
+ * version 2, as published by the Free Software Foundation.
+ *
+ * This program is distributed in the hope it will be useful, but WITHOUT
+ * ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
+ * FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General Public License for
+ * more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ */
+#pragma once
+#include <stratosphere/lm/detail/lm_detail_log_types.hpp>
+
+namespace ams::lm::detail {
+
+    class LogPacketTransmitterBase {
+        NON_COPYABLE(LogPacketTransmitterBase);
+        NON_MOVEABLE(LogPacketTransmitterBase);
+        public:
+            using FlushFunction = bool (*)(const void*, size_t);
+        private:
+            LogPacketHeader *header;
+            u8 *log_buffer_start;
+            u8 *log_buffer_end;
+            u8 *log_buffer_payload_start;
+            u8 *log_buffer_payload_current;
+            bool is_tail;
+            FlushFunction flush_function;
+        public:
+            LogPacketTransmitterBase(u8 *log_buffer, size_t log_buffer_size, FlushFunction flush_fn, diag::LogSeverity severity, bool verbosity, u64 process_id, bool head, bool tail);
+
+            ~LogPacketTransmitterBase() {
+                this->Flush(this->is_tail);
+            }
+
+            void PushData(const void *data, size_t data_size);
+            bool Flush(bool tail);
+            void PushDataChunkImpl(LogDataChunkKey key, const void *data, size_t data_size, bool is_string);
+            size_t PushUleb128(u64 value);
+            size_t GetPushableDataSize(size_t sz, size_t required_size);
+
+            inline void PushDataChunk(LogDataChunkKey key, const void *data, size_t data_size) {
+                return this->PushDataChunkImpl(key, data, data_size, false);
+            }
+
+            inline void PushDataChunk(LogDataChunkKey key, const char *str, size_t str_len) {
+                return this->PushDataChunkImpl(key, str, str_len, true);
+            }
+
+            constexpr size_t GetRemainSize() {
+                return static_cast<size_t>(this->log_buffer_end - this->log_buffer_payload_current);
+            }
+
+            static constexpr size_t GetRequiredSizeToPushUleb128(u64 value) {
```

Comment:
This seems fake/has messed up access controls. I'll refactor this myself after the detail -> impl changes are made.